### ISSUE TITLE
Nested include/excludes don't work

### ISSUE THREAD

**Issue by mikix** [restic/restic#3408]
Output of `restic version`
--------------------------
`restic 0.12.0 compiled with go1.15.8 on linux/amd64`

How did you run restic exactly?
-------------------------------

```
$ mkdir -p a/b/c
$ RESTIC_PASSWORD=foo restic -r repo init
$ RESTIC_PASSWORD=foo restic -r repo backup --exclude=a/b a a/b/c
$ RESTIC_PASSWORD=foo restic -r repo ls latest
/a
/a/b
```

What backend/server/service did you use to store the repository?
----------------------------------------------------------------
Just local.

Expected behavior
-----------------

Personally, I feel like restic *could* know that it should include `a` and `a/b/c` but exclude everything in `a/b` *except* `a/b/c`.

For example, I'd expect `a/b/some-other-folder` to not be included. But `a/b/c` to make it into the backup.

Actual behavior
---------------

Instead, it includes `a` and (curiously) `a/b`.

Note that without adding the `a/b/c` include, `a/b` is not included.

Steps to reproduce the behavior
-------------------------------
Above script.

Do you have any idea what may have caused this?
-----------------------------------------------
From quickly scanning the code, it looks like all excludes are "final" and restic does not attempt to investigate any nested include/excludes.

Do you have an idea how to solve the issue?
-------------------------------------------
I'm not familiar enough with the restic code base yet. But if this is not an interesting problem to solve for you folks, but you would accept patches for it, I can eventually give it a go.

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------
This question is cute. :)  Restic has been very good so far. I'm actually the maintainer of a backup GUI app (deja-dup) and am exploring (very rough/initial) restic support.

I really appreciate restic's `--json` argument! And may have some patches in future to issue some errors in json format, if they would be accepted.

**Comment by ctonsing** [comment]
I believe I have found a similar problem: https://forum.restic.net/t/strange-exclude-behaviour/4301

**Comment by Bananaman** [comment]
+1 this is a pretty big downside. Excluding a huge tree of junk but including a specific sub-folder that contains some vital configuration is a pretty common need in a backup program.

@MichaelEischer Since you mentioned this issue in another ticket 3 weeks ago, I am curious if you have any thoughts about Restic in this area.

**Comment by MichaelEischer** [comment]
I mainly consider this to be a feature that could be useful for some users (I've never had that problem) but I haven't thought much about it, as I'm busy with other issues/PRs.

It might be sufficient to check that excludes match complete paths / the last path components. But there are probably a few corner cases hidden somewhere.

**Comment by rafalh** [comment]
Wouldn't exclamation mark help here as described in https://restic.readthedocs.io/en/latest/040_backup.html#excluding-files ? In documentation there is an example with the home folder:
```
$HOME/*
!$HOME/Documents
!$HOME/code
!$HOME/.emacs.d
!$HOME/games
```
In case of `a/b/c` this would work, wouldn't it?
```
RESTIC_PASSWORD=foo restic -r repo backup --exclude='a/b/*' --exclude='!a/b/c' a
```

**Comment by mikix** [comment]
Yes indeed... You need to actually make sure the negation is for a `regex match` not just just any `excluded file` (which is why you used the `a/b/*` form rather than just excluding `a/b`).

This gets complicated quick if you are looking at nested paths. To exclude `a/b` but add back `a/b/c/d`, you need to exclude and add back each branch of the tree (as far as I can tell). Something like:

```
RESTIC_PASSWORD=foo restic -r repo backup --exclude='a/b/*' --exclude='!a/b/c' --exclude='a/b/c/*' --exclude='!a/b/c/d' a
```

But... it does seem to work! So that's a useful workaround indeed. Thanks rafalh!

**Comment by mikix** [comment]
Though I still think this ticket / feature request is valid - in my opinion, ideally restic would be able to know what to do when including `a` and `a/b/c/d` but excluding `a/b`. But the workaround helps a lot.

**Comment by Bananaman** [comment]
Thanks for the continued discussion. I agree with mikix, a first-class "nested include-exclude" system is needed.

The exclamation workaround seems weird, and I don't understand how it would even work. I think it only wroks because of the "asterisk" in the first example. The first asterisk basically matches every folder under Home, and then the exclamation mark re-adds those folders. But the docs explains that this trick cannot be used to include deeper folders again.

So you cannot do something like "exclude /home/foo/games/300gbgame, include /home/foo/games/300gbgame/saves".

Here's the doc excerpt again:

> "If a pattern starts with exclamation mark and matches a file that was previously matched by a regular pattern, the match is cancelled. It works similarly to gitignore, with the same limitation: once a directory is excluded, it is not possible to include files inside the directory. Here is a complete example to backup a selection of directories inside the home directory. It works by excluding any directory, then selectively add back some of them."

Their example is:

```
$HOME/*
!$HOME/Documents
!$HOME/code
!$HOME/.emacs.d
!$HOME/games
# [...]
node_modules
*~
*.o
*.lo
*.pyc
```

PS: It seems like Borg has the feature? Or does it have the same limitation of only being one-level deep? See https://www.reddit.com/r/BorgBackup/comments/v3o1gb/does_borg_support_nested_includesexcludes/

**Comment by Bananaman** [comment]
As for implementing it, I'd be happy with a flag to enable slower backups that test all patterns on every file/folder, and doesn't "short circuit/skip" anything. That would be an easy solution. Just "check every pattern on every file/folder". The only folders that would be skippable are those that definitely aren't part of any inclusion patterns to begin with.

**Comment by MichaelEischer** [comment]
Hmm, I think the current behavior might be the result of allowing patterns to match any part of the filepath and not requiring it to also match the last part. That is current an exclude for `a/b` matches both `a/b` and `a/b/c`. With the restriction that the match must also include the last part of the filepath, `a/b/c` would no longer be matched and everything should work as expected.

I haven't thought about this too much so far, so there might be some pretty large pitfalls here. But at least at first glance, it appears like it could work without causing (lots of) problems.

**Comment by Bananaman** [comment]
@MichaelEischer A naive implementation which tests all files (doesn't skip sub-folders/sub-files if a parent is excluded) would be the easiest.

A smarter implementation would be one that splits the regex or patterns on the "/" separator, and uses it to determine that `a/b/c` means that we want to explore deeper into `a` and into `a/b`, even if `a` itself is excluded. However, splitting a regex is a very difficult thing. One way to achieve it would be to use a regex engine that

**Comment by Bananaman** [comment]
@MichaelEischer A naive implementation which tests all files (doesn't skip sub-folders/sub-files if a parent is excluded) would be the easiest.

A smarter implementation would be one that splits the regex or patterns on the "/" separator, and uses it to determine that `a/b/c` means that we want to explore deeper into `a` and into `a/b`, even if `a` itself is excluded. However, splitting a regex is a very difficult thing. One way to achieve it would be to use a regex engine that allows feedback about where a match stops. Then that can be used to determine if the parent folder should be investigated deeper.

Example:

Pattern is: `^a/b/c`.

Folder to test is: `a/b`

Regex engine says: "regex matched until character 3 (b)".

Then do a check if `a/b` (the path to check) has reached the end of the word or a directory separator. It has reached the end, so we now know: "c" **must** be a sub-location under `a/b`. Therefore, it's time to investigate deeper and recurse into `a/b` to test the files underneath it.

Again, to sum it up since my explanation is probably a bit messy:

- When the regex engine tells you where in the input string (tested path) the match ended...
- ...we can use that offset in the input string (tested path) to figure out if that's a folder that would be a **parent** of the inclusion pattern.
- Example again: If the regex  `^a/b/c` matches until the 3rd char of `a/b`, we know that it's worth looking into that folder because the regex most likely refers to a sub-file underneath that folder. If it instead just matches until the 3rd char of `a/boo`, then we can see that "boo" in the input string is a longer name (the char immediately after where the regex gave up isn't a directory separator or end of string), so we know that "boo" is not worth looking deeper into.

There may be caveats that I haven't thought about (this was just a quick idea), and I am not sure how well it would work in practice. But if there's a way to know where a regex match gave up, it would be possible to infer whether it's worth scanning deeper into an otherwise-excluded folder hierarchy.

**Comment by MichaelEischer** [comment]
> A smarter implementation would be one that splits the regex or patterns on the "/" separator, and uses it to determine that `a/b/c` means that we want to explore deeper into `a` and into `a/b`, even if `a` itself is excluded. However, splitting a regex is a very difficult thing. One way to achieve it would be to use a regex engine that allows feedback about where a match stops. Then that can be used to determine if the parent folder should be investigated deeper.

The exclude patterns are glob patterns, not regexes! We already split the patterns at the directory separator, as that is necessary to properly implement `**` and `*`.

**Comment by MichaelEischer** [comment]
> A naive implementation which tests all files (doesn't skip sub-folders/sub-files if a parent is excluded) would be the easiest.

I was mostly wondering whether we'd end up with unexpected changes in the high-level behavior and actually not so much about the implementation (which would be the step afterwards).

**Comment by alexandru0-dev** [comment]
any update on the topic?

**Comment by dgasaway** [comment]
Since someone brought up borg, it does have separate patterns of exclusions that terminate directory traversal and exclusions that do not.  However, borg has a very different last-match-wins approach.  I think rustic's `--glob` does this, but without an exclude-but-traverse option.
